multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,181 @@
+      // Write the dataframe to the specified output folder as csv file
+      .write.mode(SaveMode.Overwrite).option("header", "true").csv(outputDirectory)
+  }
+}
+
+/*
+ * A custom Executor that contains the logic to perform verify and then multipass geocode
+ */
+class CustomExecutor extends AddressingExecutor {
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  // set preference to return all fields and Match Mode to Relaxed
+  private val relaxedGeocoderPreferences: Preferences = new PreferencesBuilder()
+    .withReturnAllInfo(true)
+    .withMatchMode("RELAXED")
+    .build()
+
+  override def execute(input: RequestInput, preferences: Option[Preferences], addressing: Addressing): Response = {
+
+    val verifyResponse = addressing.verify(input.requestAddress(), preferences.orNull)
+
+    var multiLineRequest: RequestAddress = input.requestAddress()
+
+    var singleLine = input.requestAddress().getStreet + " " +
+      input.requestAddress().getCity + " " +
+      input.requestAddress().getAdmin1 + " " +
+      input.requestAddress().getPostalCode
+
+    if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      val verifyAddress = verifyResponse.getResults.get(0).getAddress
+      //  create multiLine request from verify Response
+      multiLineRequest = new RequestAddress()
+      multiLineRequest.setStreet(verifyAddress.getStreet)
+      multiLineRequest.setAddressNumber(verifyAddress.getAddressNumber)
+      multiLineRequest.setAdmin1(verifyAddress.getAdmin1.getLongName)
+      multiLineRequest.setAdmin2(verifyAddress.getAdmin2.getLongName)
+      multiLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+      multiLineRequest.setCity(verifyAddress.getCity.getLongName)
+      multiLineRequest.setPostalCode(verifyAddress.getPostalCode)
+
+      // create singleline request from verify Response
+      singleLine = verifyAddress.getFormattedAddress + " " + verifyAddress.getCountry.getIsoAlpha3Code
+    }
+
+    val multiLineResponse = addressing.geocode(multiLineRequest, relaxedGeocoderPreferences)
+
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults) && isAddressLevelMatch(multiLineResponse)) {
+      return multiLineResponse
+    }
+
+    val singleLineRequest: RequestAddress = new RequestAddress()
+    singleLineRequest.setAddressLines(Collections.singletonList(singleLine))
+
+    val singleLineResponse = addressing.geocode(singleLineRequest, relaxedGeocoderPreferences)
+
+    if (CollectionUtils.isNotEmpty(singleLineResponse.getResults) && isAddressLevelMatch(multiLineResponse)) {
+      return singleLineResponse
+    }
+
+    if (CollectionUtils.isNotEmpty(singleLineResponse.getResults) && CollectionUtils.isNotEmpty(multiLineResponse.getResults) &&
+      multiLineResponse.getResults.get(0).getScore > singleLineResponse.getResults.get(0).getScore) {
```

Comment:
make this >= so that multiLine is default over single line in case of tie

Reply:
Actually, there's a few problems with this logic (even combined with below statement).  Consider a case where we have gotten a result from multi-line with score of 88 but no response from single line.  Unlikely - but i theoretically possible.  In that case, with this logic, we'd be returning the singleLine response and missing our proper candidate.
In the case where both had responses but single line was better - we'd get the right answer with this code, but its still a touch tough to follow and i can see if some changed that last disconnected line the logic could end up flawed.
I feel like it might make sense to do something like:
`
if(isNotEmpty(multiLineResponse.getResults)){
  // accept single line only if we have a result and its score is better than multi line
  if(isNotEmpty(singleLine.getResults) && singleLineResponse.getResults.get(0).getScore > multiLineResponse.getResults.get(0).getScore) {
    return singleLineResponse
  } 
  // accept multi line by default
  return multiLineResponse
} else {
  // no multi line result - return single line
  return singleLineResponse
}
`

Reply:
I think the scenario which is mentioned above will work in the current logic as well( the execution flow will not go in if block of score check because of null single line response then It will move to last if block and as single line response is null it will return multi line response ).

Although, I also think the code which you shared is more concise and efficient. So, I will update the change. Thanks